File: src/main/java/com/ninja/ieumnori/ieumnoriserver/common/domain/BaseEntity.java
package com.ninja.ieumnori.ieumnoriserver.common.domain;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;

@MappedSuperclass
@Getter
@Setter
public abstract class BaseEntity {
    @Column(name = "is_deleted", nullable = false)
    protected boolean isDeleted = false;

    @Column(name = "created_at", nullable = false, updatable = false)
    private Long createdAt;

    @Column(name = "modified_at", nullable = false)
    private Long modifiedAt;

    @Column(name = "deleted_at")
    private Long deletedAt;

    @PrePersist
    protected void onCreate() {
        long now = System.currentTimeMillis();
        this.createdAt = now;
        this.modifiedAt = now;
    }

    @PreUpdate
    protected void onUpdate() {
        this.modifiedAt = System.currentTimeMillis();
    }
}

File: src/main/java/com/ninja/ieumnori/ieumnoriserver/common/domain/BaseEntity.kt
package com.ninja.ieumnori.ieumnoriserver.common.domain

import jakarta.persistence.Column
import jakarta.persistence.MappedSuperclass
import jakarta.persistence.PreUpdate
import java.time.LocalDateTime

@MappedSuperclass
abstract class BaseEntity (
    @Column(name = "is_deleted", nullable = false)
    var isDeleted: Boolean = false,

    @Column(name = "created_at", nullable = false, updatable = false)
    val createdAt: LocalDateTime = LocalDateTime.now(),

    @Column(name = "modified_at", nullable = false)
    var modifiedAt: LocalDateTime = LocalDateTime.now(),

    @Column(name = "deleted_at", nullable = true)
    var deletedAt: LocalDateTime? = null
) {
    @PreUpdate
    fun preUpdate() {
        modifiedAt = LocalDateTime.now()
    }
}